modules/functional-test/src/main/java/com/intuit/wasabi/tests/service/TimeStampTest.java

```diff
@@ -0,0 +1,187 @@
+ * amount of impressions are being created when using different formats of timestamps at various times.
+ */
+public class TimeStampTest extends TestBase {
+    private static final Logger LOGGER = getLogger(BatchRuleTest.class);
+    public static final String COUNT_ERROR_MESSAGE = "Impression count does not match count of timestamps";
+    public static final String TIMING_ERROR_MESSAGE = "Found impressions after expected time";
+    private final List<Experiment> validExperimentsLists = new ArrayList<>();
+    private Map<User, Assignment> assignments = new HashMap<>();
+
+    @DataProvider
+    public Object[][] sampleExperiment() {
+        String label = "timestampTest_" + System.currentTimeMillis();
+        Application application = new Application("LUA_timestamp");
+        String startTime = "2013-01-01T00:00:00+0000";
+        String endTime = TestUtils.relativeTimeString(1);
+        Double samplingPercent = 1.0;
+        Experiment experiment = new Experiment(label, application, startTime, endTime, samplingPercent);
+        return new Object[][]{
+                new Object[]{experiment}
+        };
+    }
+
+    @Test(dataProvider = "sampleExperiment")
+    public void setupExperiment(Experiment experimentData) {
+        LOGGER.debug(format("posting experiment: %s", experimentData.toString()));
+        response = apiServerConnector.doPost("/experiments", experimentData);
+        Experiment experiment = createFromJSONString(response.jsonPath().prettify());
+        validExperimentsLists.add(experiment);
+        assertThat(response.getStatusCode(), is(CREATED.getStatusCode()));
+    }
+
+    @Test(dependsOnMethods = {"setupExperiment"})
+    public void setupBucketsAndStartExperiment() {
+        // Create bucket with 100% allocation and change experiment state to running
+        String greenBucket = "{\"label\": \"green\", \"allocationPercent\": 1.0, \"isControl\": true, \"description\": \"Green buy button\"}";
+        for (Experiment experiment : validExperimentsLists) {
+            response = apiServerConnector.doPost("/experiments/" + experiment.id + "/buckets", greenBucket);
+            assertThat(response.getStatusCode(), is(CREATED.getStatusCode()));
+            response = apiServerConnector.doPut("/experiments/" + experiment.id, "{\"state\": \"RUNNING\"}");
+            assertThat(response.getStatusCode(), is(OK.getStatusCode()));
+        }
+    }
+
+    // Ensures that the correct amount of impressions are being registered with respect to the timestamp and experiment
+    @Test(dependsOnMethods = {"setupBucketsAndStartExperiment"})
+    public void gatherImpressions() throws InterruptedException {
+        // Create fake user
+        String username = "timestamp_user";
+        User outUser = createUser(username);
+
+        // Create identical timestamps in different formats
+        String prefix = "2013-09-21T09:00:00";
+        List<String> offSets = new ArrayList<>();
+        offSets.add("Z");
```

Comment:
instead of coding the possible test cases here, maybe each offset an entry in a data provider